symengine/expansion.cpp

```diff
@@ -0,0 +1,619 @@
+    piranha::rational r(i1);
+    r /= i2;
+    return r;
+}
+
+static inline piranha::rational mpqc2prat(const mpq_class& the_mpqc)
+{
+    piranha::integer i1(the_mpqc.get_num_mpz_t());
+    piranha::integer i2(the_mpqc.get_den_mpz_t());
+    piranha::rational r(i1);
+    r /= i2;
+    return r;
+}
+
+static const std::list<unsigned int>& step_list(unsigned int prec)
+{
+    static std::list<unsigned int> steps;
+    if (not steps.empty()) {
+        if (*(steps.rbegin()) == prec)
+            return steps;
+        else
+            steps.clear();
+    }
+
+    unsigned int tprec = prec;
+    while (tprec > 4) {
+        tprec = 2 + tprec/2;
+        steps.push_front(tprec);
+    }
+    steps.push_front(2);
+    steps.push_back(prec);
+    return steps;
+}
+
+pp_t series_invert(const pp_t s, const RCP<const Symbol> &var, unsigned int prec)
+{
+    if (s == 0)
+        throw std::runtime_error("series_invert: division by zero");
+    if (s == 1)
+        return pp_t{1};
+    const short ldeg = s.ldegree();
+    const piranha::rational& co = s.find_cf({ldeg});
+    const std::string &varname = var->get_name();
+    const pp_t var_p{varname};
+    pp_t p(co.pow(-1)), ss = s;
+    if (ldeg != 0)
+        ss = s * var_p.pow(-ldeg);
+    auto steps = step_list(prec);
+    for (const auto step : steps) {
+        pp_t::set_auto_truncate_degree(step, {varname});
+        p -= p*(p*ss - 1);
+    }
+    if (ldeg != 0) {
+        return p * var_p.pow(-ldeg);
+    }
+    else
+        return p;
+}
+
+pp_t series_reverse(const pp_t s, const RCP<const Symbol> &var, unsigned int prec)
+{
+    const piranha::rational& co = s.find_cf({0});
+    if (co != 0)
+        throw std::runtime_error("reversion of series with constant term");
+    const piranha::rational& a = s.find_cf({1});
+    if (a == 0)
+        throw std::runtime_error("reversion of series with zero term of degree one");
+    const std::string &varname = var->get_name();
+    pp_t r(varname), v(varname);
+    r /= a;
+    for (unsigned int i=2; i<prec; i++) {
+        pp_t::set_auto_truncate_degree(i+1, {varname});
+        pp_t sp(s.subs(varname, r));
+        r -= (v.pow(i) * sp.find_cf({i}))/a;
+    }
+    pp_t::set_auto_truncate_degree(prec, {varname});
+    return r;
+}
+
+pp_t series_nthroot(const pp_t s, int n, const RCP<const Symbol> &var, unsigned int prec)
+{
+    if (n == 0)
+        return pp_t{1};
+    if (n == 1)
+        return s;
+    if (n == -1)
+        return series_invert(s, var, prec);
+
+    piranha::rational ct(s.find_cf({0}));
+    if (ct == 0)
+        throw std::runtime_error("series_nthroot needs a constant term");
+    const std::string& varname = var->get_name();
+    mpq_class cl_rat(ct.get_mpq_view());
+    mpq_class cl_root(cl_rat);
+    bool res;
+    if (mpz_cmp_ui(cl_rat.get_den_mpz_t(), 1) == 0) {
+        // integer constant
+        res = mpz_root(cl_root.get_num_mpz_t(), cl_rat.get_num_mpz_t(), n) != 0;
```

Comment:
In a travis test `n` is `-2`, but the function signature is,
`int mpz_root (mpz_t rop, const mpz_t op, unsigned long int n)` and this is ignored silently.
We should fix #228